extensions/wikia/AdEngine/js/AdEngine2.js

```diff
@@ -60,6 +60,32 @@ define('ext.wikia.adEngine.adEngine', [
 		return providerContainer;
 	}
 
+	/**
+	 * Initialize the provider before the first use
+	 * Build the queue for fillInSlot and call initialize (if present)
+	 *
+	 * !! If initialize method is present in a provider it MUST accept a callback param
+	 * and call it back once it is initialized !!
+	 *
+	 * TODO (if useful): support error param, so initialize can cause a provider to always hop?
+	 */
+	function initializeProviderOnce(provider) {
+		if (!provider.fillInSlotQueue) {
+			provider.fillInSlotQueue = [];
+			lazyQueue.makeQueue(provider.fillInSlotQueue, function (args) {
+				provider.fillInSlot.apply(provider, args);
+			});
+
+			if (provider.initialize) {
+				// Only start flushing the fillInSlot queue once the provider is initialized
+				provider.initialize(provider.fillInSlotQueue.start);
```

Comment:
Since you mentioned flushing here I wonder how it affects our ATF implementation. It doesn't, does it? Currently we don't have a provider with `initialize()` method. Do you think we should re-implement ATF/BTF logic with this function?